src/dynamic_programming/edit_distance.rs

```diff
@@ -0,0 +1,119 @@
+//! Compute the edit distance between two strings
+
+use std::cmp::min;
+
+/// edit_distance(str_a, str_b) returns the edit distance between the two
+/// strings This edit distance is defined as being 1 point per insertion,
+/// substitution, or deletion which must be made to make the strings equal.
+///
+/// This function iterates over the bytes in the string, so it may not behave
+/// entirely as expected for non-ASCII strings.
+///
+/// # Complexity
+///
+/// - time complexity: O(nm),
+/// - space complexity: O(nm),
+///
+/// where n and m are lengths of `str_a` and `str_b`
+pub fn edit_distance(str_a: &str, str_b: &str) -> u32 {
+    // distances[i][j] = distance between a[..i] and b[..j]
+    let mut distances = vec![vec![0; str_b.len() + 1]; str_a.len() + 1];
+    // Initialize cases in which one string is empty
+    for j in 0..=str_b.len() {
+        distances[0][j] = j as u32;
+    }
+    for (i, item) in distances.iter_mut().enumerate() {
+        item[0] = i as u32;
+    }
+    for i in 1..=str_a.len() {
+        for j in 1..=str_b.len() {
+            distances[i][j] = min(distances[i - 1][j] + 1, distances[i][j - 1] + 1);
+            if str_a.as_bytes()[i - 1] == str_b.as_bytes()[j - 1] {
+                distances[i][j] = min(distances[i][j], distances[i - 1][j - 1]);
+            } else {
+                distances[i][j] = min(distances[i][j], distances[i - 1][j - 1] + 1);
+            }
+        }
+    }
+    distances[str_a.len()][str_b.len()]
+}
+
+/// The space efficient version of the above algorithm.
+///
+/// Instead of storing the `m * n` matrix expicitly, only one row (of length `n`) is stored.
+/// It keeps overwriting itself based on its previous values with the help of two scalars,
+/// gradually reaching the last row. Then, the score is `matrix[n]`.
+///
+/// # Complexity
+///
+/// - time complexity: O(nm),
+/// - space complexity: O(n),
+///
+/// where n and m are lengths of `str_a` and `str_b`
+pub fn edit_distance_se(str_a: &str, str_b: &str) -> u32 {
+    let (str_a, str_b) = (str_a.as_bytes(), str_b.as_bytes());
+    let (m, n) = (str_a.len(), str_b.len());
+    let mut distances: Vec<u32> = vec![0; n + 1]; // the dynamic programming matrix (only 1 row stored)
```

Comment:
Comments are for humans to be read. Let's always capitalize them and end them in `.`.

```suggestion
    let mut distances: Vec<u32> = vec![0; n + 1]; // The dynamic programming matrix (only 1 row stored).
```